### PR TITLE
As a developer, I can create Gin project from command

### LINKED ISSUE #8: Create the generator for the application
## Why
Create the generator with the CLI to create the app and replace all the skeleton values with the correct value (such as database name or docker container name).

Note:
After some researched, I prefer using [Cobra](https://github.com/spf13/cobra) to generate the CLI command as it is popular and used by big organizations such as Kubernetes, Hugo, Github and view more [here](https://github.com/spf13/cobra/blob/master/projects_using_cobra.md).
Also, maybe we can use [cookiecutter](https://github.com/lacion/cookiecutter-golang) to fill the skeleton values.

## Acceptance Criteria
- Generated app can run properly.

### PR DESCRIPTION
**Pull request by gutakk**
## What happened 👀
Resolve #8 
Implemented generator to create Gin project from template.
 
## Insight 📝
- Implemented `create` command using [Cobra](https://github.com/spf13/cobra) to wrap [cookiecutter](https://github.com/lacion/cookiecutter-golang) command.

**Testing**
`go get github.com/nimblehq/gin-templates` and `gin-template create` command will use source code from default git branch (which is `develop` in this case) so the instruction in the README won't work until we merge this to develop branch.

There are 2 solutions to test this feature.
1. Change the default branch to this branch during testing (with this solution, you can follow the instruction in README)
2. Follow the following steps
    - Clone this repo and checkout to this branch
    - Edit the code in `create.go` at line 31 from `https://github.com/nimblehq/gin-templates` to `https://github.com/gutakk/gut-cookiecutter`
    - Change directory to the repo.
    - Run `go build -i -o $GOPATH/bin/nimble-gin`
    - Run `nimble-gin create`
 
## Proof Of Work 📹
- Followed the second solution
losbpG3fLR

### CODE REVIEW

File: cmd/create.go [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+	Use:   "create",
+	Short: "Create the gin project with your own app name.",
+	Run: func(cmd *cobra.Command, args []string) {
+		shCmd := exec.Command("cookiecutter", "https://github.com/nimblehq/gin-templates")
+		shCmd.Stdout = os.Stdout
+		shCmd.Stdin = os.Stdin
+
+		err := shCmd.Run()
+		if err != nil {
+			log.Fatal("Create unsuccessfully: ", err)
```
Comment (junan) [inline cmd/create.go]: >Create unsuccessfully:

How about use `Create failed:`?

File: cmd/create.go [hunk 2]
```diff
@@ -0,0 +1,46 @@
+/*
+Copyright © 2014-2020 Nimble
```
Comment (junan) [inline cmd/create.go]:
```suggestion
Copyright © 2014-2021 Nimble
```

File: cmd/root.go [hunk 3]
```diff
@@ -0,0 +1,38 @@
+/*
+Copyright © 2014-2020 Nimble
```
Comment (junan) [inline cmd/root.go]:
```suggestion
Copyright © 2014-2021 Nimble
```

File: main.go [hunk 4]
```diff
@@ -0,0 +1,22 @@
+/*
+Copyright © 2014-2020 Nimble
```
Comment (junan) [inline main.go]:
```suggestion
Copyright © 2014-2021 Nimble
```

Comment (gutakk) [reply main.go]: Fixed in 4b74b60b785dc6e389b0ee1f9ae891de84fcbad6

File: cmd/root.go (hunk 3 above)
Comment (gutakk) [reply cmd/root.go]: Fixed in 4b74b60b785dc6e389b0ee1f9ae891de84fcbad6

File: cmd/create.go (hunk 2 above)
Comment (gutakk) [reply cmd/create.go]: Fixed in 4b74b60b785dc6e389b0ee1f9ae891de84fcbad6

File: cmd/create.go (hunk 1 above)
Comment (gutakk) [reply cmd/create.go]: How about `Gin project created unsuccessfully: ...`? This will consistent with the successful case.

**Review by carryall — APPROVED** [review]

File: README.md [hunk 5]
```diff
@@ -12,12 +12,25 @@ Our templates offer a rich boilerplate to jump start Go Gin-based application de
...
+  go get github.com/nimblehq/gin-templates
+  ```
 
-`git clone git@github.com:nimblehq/git-template.git`
+- Run create the project command.
+  ```
+  gin-templates create
+  ```
 
+- Follow the instruction and the app should be created.
```
Comment (olivierobert) [inline README.md]: Line break needed and rephrasing needed:
```suggestion
- Follow the instructions in the terminal.
- You new application is created 🎉 .

```

File: cmd/root.go [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+import (
+	"fmt"
+	"os"
+
+	"github.com/spf13/cobra"
+)
+
+// rootCmd represents the base command when called without any subcommands
+var rootCmd = &cobra.Command{
+	Use:  "gin-templates",
```
Comment (olivierobert) [inline cmd/root.go]: Just like for [Elixir templates](https://github.com/nimblehq/elixir-templates), we should prefix it with `nimble`:

 ```suggestion
	Use:  "nimble-gin",
```

We could also omit templates to sound cooler 😎

File: {{cookiecutter.app_name}}/go.mod [hunk 7]
```diff
@@ -0,0 +1,3 @@
+module github.com/nimblehq/{{cookiecutter.app_name}}
```
Comment (olivierobert) [inline {{cookiecutter.app_name}}/go.mod:1]: To keep the template code away from the `gin-templates` app codebase, I recommend to have a directory named `template` and place all code for the app template in there.

For now, there would be only `go.mod`.

With this, please check if we can avoid having the directory `{{cookiecutter.app_name}}` by creating this by ourselves and then adding all files from the template directory. It's not great to have a directory named with braces 🥴

File: README.md (hunk 5 above)
Comment (gutakk) [reply README.md]: Fixed in abd9be910833e2fd834ea6f9a0349a1cbbd4b5ba

File: cmd/root.go (hunk 6 above)
Comment (gutakk) [reply cmd/root.go]: Fixed in 0264db2eb9e3b9029c11c0dff523e88a7968722e

File: {{cookiecutter.app_name}}/go.mod (hunk 7 above)
Comment (gutakk) [reply {{cookiecutter.app_name}}/go.mod:1]: After an hour researched, cookiecutter doesn't support what you mentioned atm. I opened an issue to ask them about what we need and waiting for their reply now.

We can do like this atm. What do you think?
```
.
├── .github/
├── cmd/
├── template/
    ├── {{cookiecutter.app_name}
        ├── go.mod
    ├── cookiecutter.json
├── main.go
```

File: cmd/root.go (hunk 3 above)
Comment (olivierobert) [reply cmd/root.go]: For this, please follow the updated [Git template repository](https://github.com/nimblehq/git-template): https://github.com/nimblehq/git-template/blob/main/LICENSE#L1

File: cmd/create.go (hunk 2 above)
Comment (olivierobert) [reply cmd/create.go]: Same as the comment below, instead follow the updated Git template.

File: {{cookiecutter.app_name}}/go.mod (hunk 7 above)
Comment (olivierobert) [reply {{cookiecutter.app_name}}/go.mod:1]: Too bad. Then that's a tradeoff we would need to consider when organizing the codebase

**Comment by gutakk** [comment]
After discussed with @olivierobert, We will use cookiecutter for release 1.0.0 and maybe we will find another way in release 2.0.0

File: cmd/root.go (hunk 3 above)
Comment (gutakk) [reply cmd/root.go]: Updated in 22d2254ae3f8e73cbab74d5725e62ad7a2a407b0

**Comment by gutakk** [comment]
@olivierobert Updated copyright in 6c685913c275dfcdb2bdcf8f144403958ea3d725 22d2254ae3f8e73cbab74d5725e62ad7a2a407b0

**Review by olivierobert — APPROVED** [review]

**Pull request merged**